### PR TITLE
fix some bugs in spine webassembly

### PR DESCRIPTION
**Pull request by zhakesi**

### CODE REVIEW

**Review by SantyWang — APPROVED** [review]

**Comment by SantyWang** [comment]
@cocos-robot run test cases

File: native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp [hunk 1]
```diff
@@ -50,4 +50,9 @@ AttachmentVertices::~AttachmentVertices() {
     if (_texture) _texture->release();
 }
 
+AttachmentVertices *AttachmentVertices::copy() {
+    AttachmentVertices *atv = new AttachmentVertices(nullptr, _triangles->vertCount, _triangles->indices, _triangles->indexCount);
+    return atv;
+}
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp:56]: Where to delete the memeory?

File: cocos/spine/lib/spine-core.d.ts [hunk 2]
```diff
@@ -1277,7 +1277,7 @@ declare namespace spine {
     class SkeletonInstance {
         initSkeleton(data: SkeletonData);
         getAnimationState();
-        setAnimation(trackIndex: number, name: string, loop: boolean);
+        setAnimation(trackIndex: number, name: string, loop: boolean): spine.TrackEntry;
```
Comment (dumganhar) [inline cocos/spine/lib/spine-core.d.ts]: Should the return value be `spine.TrackEntry | null`?

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 3]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+    if (!_skeleton) return;
+    auto slot = _skeleton->findSlot(slotName.c_str());
+    if (!slot) return;
+    auto attachment = slot->getAttachment();
+    if (!attachment) return;
+    auto width = tex2d->getWidth();
+    auto height = tex2d->getHeight();
+
+    if (createAttachment) {
+        attachment = attachment->copy();
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp:1056]: Here to create new attachment. But the above logic will return if attachment is null. The logic seems wired.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 4]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        slot->setAttachment(attachment);
+    }
+    AttachmentVertices *attachmentVertices = nullptr;
+    if (attachment->getRTTI().isExactly(spine::RegionAttachment::rtti)) {
+        auto region = (RegionAttachment *)attachment;
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: It is better to use static_cast.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 5]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        region->setRegionWidth(width);
+        region->setRegionHeight(height);
+        region->setRegionOriginalWidth(width);
+        region->setRegionOriginalHeight(height);
+        region->setWidth(width);
+        region->setHeight(height);
+        region->setUVs(0, 0, 1.0f, 1.0f, false);
+        region->updateOffset();
+        attachmentVertices = (AttachmentVertices *)region->getRendererObject();
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: The same as above.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 6]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        if (createAttachment) {
+            attachmentVertices = attachmentVertices->copy();
+            region->setRendererObject(attachmentVertices);
+        }
+        V3F_T2F_C4B *vertices = attachmentVertices->_triangles->verts;
+        for (int i = 0, ii = 0; i < 4; ++i, ii += 2) {
+            vertices[i].texCoord.u = region->getUVs()[ii];
+            vertices[i].texCoord.v = region->getUVs()[ii + 1];
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: May be it is better to cache region->getUVs() before loop.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 7]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        }
+    } else if (attachment->getRTTI().isExactly(spine::MeshAttachment::rtti)) {
+        auto mesh = (MeshAttachment *)attachment;
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: It is better to use static_cast

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 8]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        mesh->setWidth(width);
+        mesh->setHeight(height);
+        mesh->setRegionU(0);
+        mesh->setRegionV(0);
+        mesh->setRegionU2(1.0f);
+        mesh->setRegionV2(1.0f);
+        mesh->setRegionRotate(true);
+        mesh->setRegionDegrees(0);
+        mesh->updateUVs();
+        attachmentVertices = (AttachmentVertices *)mesh->getRendererObject();
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: The same as above.

File: cocos/spine/skeleton.ts [hunk 9]
```diff
@@ -815,9 +882,16 @@ export class Skeleton extends UIRenderer {
     /**
      * @engineInternal
      */
-    public requestDrawData (material: Material, indexOffset: number, indexCount: number) {
+    public requestDrawData (material: Material, texureID: number, indexOffset: number, indexCount: number) {
         const draw = this._drawList.add();
         draw.material = material;
+        if (texureID === 0) {
+            draw.texture = this._textures[0];
```
Comment (dumganhar) [inline cocos/spine/skeleton.ts:889]: this._textures is an array now, when will the `index > 0` be used? Do we only use the first element now?

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 10]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        if (createAttachment) {
+            attachmentVertices = attachmentVertices->copy();
+            mesh->setRendererObject(attachmentVertices);
+        }
+        V3F_T2F_C4B *vertices = attachmentVertices->_triangles->verts;
+        for (size_t i = 0, ii = 0, nn = mesh->getWorldVerticesLength(); ii < nn; ++i, ii += 2) {
+            vertices[i].texCoord.u = mesh->getUVs()[ii];
+            vertices[i].texCoord.v = mesh->getUVs()[ii + 1];
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: May be it is better to cache region->getUVs() before loop.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 11]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        if (createAttachment) {
+            attachmentVertices = attachmentVertices->copy();
+            region->setRendererObject(attachmentVertices);
+        }
+        V3F_T2F_C4B *vertices = attachmentVertices->_triangles->verts;
+        for (int i = 0, ii = 0; i < 4; ++i, ii += 2) {
+            vertices[i].texCoord.u = region->getUVs()[ii];
+            vertices[i].texCoord.v = region->getUVs()[ii + 1];
+        }
+    } else if (attachment->getRTTI().isExactly(spine::MeshAttachment::rtti)) {
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp:1081]: It seems the logic is more or less the same as `if`, can combine them?

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 12]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        }
+        V3F_T2F_C4B *vertices = attachmentVertices->_triangles->verts;
+        for (size_t i = 0, ii = 0, nn = mesh->getWorldVerticesLength(); ii < nn; ++i, ii += 2) {
+            vertices[i].texCoord.u = mesh->getUVs()[ii];
+            vertices[i].texCoord.v = mesh->getUVs()[ii + 1];
+        }
+    }
+    if (!attachmentVertices) return;
+    middleware::Texture2D* middlewareTexture = nullptr;
+    for (auto it = _slotTextureSet.begin(); it != _slotTextureSet.end(); it++) {
```
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: I think can use for in loop here.

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp [hunk 13]
```diff
@@ -1041,3 +1042,82 @@ cc::Material *SkeletonRenderer::requestMaterial(uint16_t blendSrc, uint16_t blen
...
+        mesh->setRegionV(0);
+        mesh->setRegionU2(1.0f);
+        mesh->setRegionV2(1.0f);
+        mesh->setRegionRotate(true);
+        mesh->setRegionDegrees(0);
+        mesh->updateUVs();
+        attachmentVertices = (AttachmentVertices *)mesh->getRendererObject();
+        if (createAttachment) {
+            attachmentVertices = attachmentVertices->copy();
+            mesh->setRendererObject(attachmentVertices);
```
Comment (dumganhar) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp:1099]: When will the `copied` attachmentVertices released?

File: native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp (hunk 12 above)
Comment (minggo) [inline native/cocos/editor-support/spine-creator-support/SkeletonRenderer.cpp]: I think can use for in loop here.

File: native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp [hunk 14]
```diff
@@ -339,14 +340,20 @@ void SpineSkeletonInstance::collectMeshData() {
         // record debug shape info        
         if (_userData.debugMode) {
             SpineDebugShape debugShape;
-            debugShape.type = debugShapeType;
+            debugShape.type = (uint32_t)debugShapeType;
```
Comment (minggo) [inline native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp]: It is better to use static_cast.

File: native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp [hunk 15]
```diff
@@ -339,14 +340,20 @@ void SpineSkeletonInstance::collectMeshData() {
...
             debugShape.vOffset = _model->vCount;
             debugShape.vCount = currMesh.vCount;
             debugShape.iOffset = _model->iCount;
             debugShape.iCount = currMesh.iCount;
             _debugShapes.push_back(debugShape);
         }
 
+        currMesh.blendMode = (uint32_t)slot->getData().getBlendMode();
```
Comment (minggo) [inline native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp]: The same as above.

File: native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp [hunk 16]
```diff
@@ -463,4 +470,84 @@ void SpineSkeletonInstance::onAnimationStateEvent(TrackEntry *entry, EventType t
...
+
+void SpineSkeletonInstance::resizeSlotRegion(const std::string& slotName, uint32_t width, uint32_t height, bool createNew) {
+    if (!_skeleton) return;
+    auto slot = _skeleton->findSlot(slotName.c_str());
+    if (!slot) return;
+    auto attachment = slot->getAttachment();
+    if (!attachment) return;
+    if (createNew) {
+        attachment = attachment->copy();
+        slot->setAttachment(attachment);
```
Comment (dumganhar) [inline native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp:483]: When will be the `copied` attachment released?

File: native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp [hunk 17]
```diff
@@ -463,4 +470,84 @@ void SpineSkeletonInstance::onAnimationStateEvent(TrackEntry *entry, EventType t
...
+        region->setRegionOriginalWidth(width);
+        region->setRegionOriginalHeight(height);
+        region->setWidth(width);
+        region->setHeight(height);
+        region->setUVs(0, 0, 1.0f, 1.0f, false);
+        region->updateOffset();
+        auto attachmentVertices = (AttachmentVertices*)region->getRendererObject();
+        if (createNew) {
+            attachmentVertices = attachmentVertices->copy();
+            region->setRendererObject(attachmentVertices);
```
Comment (dumganhar) [inline native/cocos/editor-support/spine-wasm/spine-skeleton-instance.cpp:498]: When will be the `copied` attachmentVertices released?

File: native/cocos/editor-support/spine-wasm/spine-type-export.cpp [hunk 18]
```diff
@@ -10,19 +10,18 @@ using namespace emscripten;
...
 namespace {
-
-std::string convertSPString2std(const spine::String& str) {
+std::string STRING_SP2STD(const spine::String& str) {
     std::string stdStr(str.buffer(), str.length());
     return stdStr;
 }
 
-const spine::String convertStd2SPString(std::string& str) {
+const spine::String STRING_STD2SP(std::string& str) {
```
Comment (dumganhar) [inline native/cocos/editor-support/spine-wasm/spine-type-export.cpp:18]: Why to rename the function to all capital?

File: cocos/spine/skeleton.ts (hunk 9 above)
Comment (zhakesi) [reply cocos/spine/skeleton.ts:889]: yes

File: native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp (hunk 1 above)
Comment (zhakesi) [reply native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp:56]: 统一解释下这里的内存问题：
AttachmentVertices 是多个组件共享的。因此在接口上层提供了createNew的选项，用于避免在多个共享数据的组件时，更改一个组件对其他组件产生影响。在createNew 为false时，在原本的数据上做修改即可，不用创建新数据。
数据会在组件销毁时释放。

这里是把释放需要新建数据的选择交给了用户。

Comment (minggo) [reply native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp:56]: Ok.

File: cocos/spine/skeleton.ts (hunk 9 above)
Comment (dumganhar) [reply cocos/spine/skeleton.ts:889]: Is it possible to use more than 1 texture in the future?

**Review by minggo — APPROVED** [review]

**Comment by minggo** [comment]
As CI doesn't check wasm lib. And previous CI passed, so merge it.

File: native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp (hunk 1 above)
Comment (dumganhar) [reply native/cocos/editor-support/spine-creator-support/AttachmentVertices.cpp:56]: 用户需要调用哪个接口去释放？

**Pull request merged**